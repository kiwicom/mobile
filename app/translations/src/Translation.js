// @flow

import * as React from 'react';
import { NativeModules } from 'react-native';
import { Text } from '@kiwicom/react-native-app-shared';

import DefaultVocabulary, { type TranslationKeys } from './DefaultVocabulary';
import CaseTransform, {
  type SupportedTransformationsType,
} from './transformations/CaseTransform';

type Props =
  | {|
      id: TranslationKeys,
      values?: Object,
      textTransform?: SupportedTransformationsType,
    |}
  | {|
      passThrough: ?string | ?number, // may be undefined because of IDX output
      textTransform?: SupportedTransformationsType,
    |};

/**
 * Regular translation using registered keys:
 *
 * ```js
 * <Translation id="SingleHotel.Rating.Stars" />
 * ```
 *
 * So called "pass through" translation (this props is not being translated
 * but passed through the component).
 *
 * ```js
 * <Translation passThrough="★★★★★" />
 * ```
 *
 * It's possible to apply text transformations in both cases:
 *
 * ```js
 * <Translation passThrough="★★★★★" textTransform="uppercase" />
 * <Translation id="SingleHotel.Rating.Stars" textTransform="uppercase" />
 * ```
 */
export default class Translation extends React.Component<Props> {
  // use this property to highlight translations for screenshoting or debugging
  highlightTranslations = false;

  replaceValues = (text: string, values?: Object): string => {
    const VARIABLE_REGEX = /__([0-9]+_)?([a-zA-Z-_]+)__/g;
    return text.replace(
      VARIABLE_REGEX,
      (match, variableIndex, variableName) => {
        // 'variableIndex' is currently being ignored (we don't need it)
        const value = values && values[variableName];
        return value !== undefined ? value : '?';
      },
    );
  };

  render = () => {
    const containerStyle = this.highlightTranslations
      ? {
          backgroundColor: this.props.id ? 'orange' : 'pink',
        }
      : undefined;

    if (this.props.id === undefined) {
      // ID prop is missing so let's assume it's "pass through" translation
      // (property passThrough may be undefined)
      return (
        <Text style={containerStyle}>
          {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
          {CaseTransform(
            String(this.props.passThrough),
            this.props.textTransform,
          )}
        </Text>
      );
    }

    const key = this.props.id;
    let translatedString = NativeModules.RNTranslationManager.translate(key);

    if (translatedString === key) {
      // fallback to our dummy vocabulary because native code didn't provide translation
      translatedString = DefaultVocabulary[key];
    }

    if (translatedString === undefined) {
      // Everything failed - native translation even our fallback. This usually
      // means we are using key that does not exist. In this case the only thing
      // we can actually do is to use original key as a text.
      translatedString = key;
    }

    const values = this.props.values || {};
    return (
      <Text style={containerStyle}>
        {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
        {CaseTransform(
          this.replaceValues(translatedString, values),
          this.props.textTransform,
        )}
      </Text>
    );
  };
}
