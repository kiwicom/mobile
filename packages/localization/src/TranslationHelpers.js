//@flow

import DefaultVocabulary, {
  type TranslationKeys,
  type TranslationKeysObject,
} from './DefaultVocabulary';

export const getTranslation = (
  translatedString: ?string,
  translationId: TranslationKeys,
  vocabulary: TranslationKeysObject = DefaultVocabulary,
): string => {
  const nativeKey = `mobile.${translationId}`;
  const defaultTranslation = vocabulary[translationId];

  if (translatedString === nativeKey) {
    // fallback to our dummy vocabulary because native code didn't provide translation
    return defaultTranslation;
  }
  if (translatedString == null) {
    // Everything failed - native translation even our fallback. This usually
    // means we are using key that does not exist. In this case the only thing
    // we can actually do is to use original key as a text.
    return translationId;
  }

  // Translated strings from native come under the form
  // %1$@, so we lose information about the variables.
  // We add this information back in:
  try {
    const NATIVE_TRANSLATION_PARAM_REGEX = /%([0-9]+)\$@/g;
    const compatibleTranslatedString = translatedString.replace(
      NATIVE_TRANSLATION_PARAM_REGEX,
      (match, variableIndex) => {
        const ORIGINAL_PARAM_REGEX = new RegExp(
          `__(${variableIndex}_)?([a-zA-Z-_]+)__`,
          'g',
        );
        const originalParamMatches = defaultTranslation.match(
          ORIGINAL_PARAM_REGEX,
        );

        if (originalParamMatches == null || originalParamMatches.length !== 1) {
          throw new Error(
            `There was a problem with translationId ${translationId}`,
          );
        }

        return originalParamMatches[0];
      },
    );
    return compatibleTranslatedString;
  } catch (e) {
    return translatedString;
  }
};

export function replaceValues(text: string, values?: Object): string {
  const VARIABLE_REGEX = /__([0-9]+_)?([a-zA-Z-_]+)__/g;
  return text.replace(VARIABLE_REGEX, (match, variableIndex, variableName) => {
    // 'variableIndex' is currently being ignored (we don't need it)
    const value = values && values[variableName];
    return value !== undefined ? value : '?';
  });
}
