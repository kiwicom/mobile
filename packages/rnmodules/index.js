// @flow

export { translate } from './translation/TranslationManager';
export { formatAmount } from './currency/CurrencyFormatter';
