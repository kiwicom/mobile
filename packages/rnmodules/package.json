{
  "name": "@kiwicom/rnmodules",
  "version": "0.1.31",
  "private": true,
  "main": "index.js",
  "peerDependencies": {
    "react-native": "^"
  }
}