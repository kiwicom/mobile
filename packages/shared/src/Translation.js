// @flow

import * as React from 'react';
import {
  CaseTransform,
  getTranslation,
  type SupportedTransformationsType,
  type TranslationKeysType,
} from '@kiwicom/mobile-localization';

import Text from './Text';

type CommonProps = {|
  testID?: string,
  textTransform?: SupportedTransformationsType,
|};

type Props =
  | {|
      id: TranslationKeysType,
      values?: Object,
      ...CommonProps,
    |}
  | {|
      passThrough: ?string | ?number, // may be undefined because of optional chaining
      ...CommonProps,
    |};

/**
 * Regular translation using registered keys:
 *
 * ```js
 * <Translation id="SingleHotel.Rating.Stars" />
 * ```
 *
 * So called "pass through" translation (this props is not being translated
 * but passed through the component).
 *
 * ```js
 * <Translation passThrough="★★★★★" />
 * ```
 *
 * It's possible to apply text transformations in both cases:
 *
 * ```js
 * <Translation passThrough="★★★★★" textTransform="uppercase" />
 * <Translation id="SingleHotel.Rating.Stars" textTransform="uppercase" />
 * ```
 */

// use this property to highlight translations for screenshoting or debugging
const highlightTranslations = false;
export default function Translation(props: Props) {
  const containerStyle = highlightTranslations
    ? {
        backgroundColor: props.id ? 'orange' : 'pink',
      }
    : undefined;
  if (props.id === undefined && props.passThrough != null) {
    return (
      <Text style={containerStyle}>
        {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
        {CaseTransform(String(props.passThrough), props.textTransform)}
      </Text>
    );
  } else if (props.id === undefined) {
    return null;
  }

  const values = props.values || {};
  return (
    <Text style={containerStyle}>
      {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
      {CaseTransform(getTranslation(props.id, values), props.textTransform)}
    </Text>
  );
}
