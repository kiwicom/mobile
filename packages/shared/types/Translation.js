// @flow

import * as React from 'react';

import Translation from '../src/Translation';
import TranslationFragment from '../src/TranslationFragment';

export type TranslationType =
  | React.Element<typeof Translation>
  | React.Element<typeof TranslationFragment>;
