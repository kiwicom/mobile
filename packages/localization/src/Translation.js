// @flow

import * as React from 'react';
import { Text } from '@kiwicom/mobile-shared';
import { translate } from '@kiwicom/rnmodules';

import vocabulary, { type TranslationKeys } from './DefaultVocabulary';
import CaseTransform, {
  type SupportedTransformationsType,
} from './transformations/CaseTransform';
import { getTranslation, replaceValues } from './TranslationHelpers';

// Preload translations
Object.keys(vocabulary).forEach(async key => {
  const nativeKey = 'mobile.' + key;

  const translatedString = await translate(nativeKey);
  if (translatedString != null) {
    vocabulary[key] = getTranslation(translatedString, key);
  }
});

type CommonProps = {|
  testID?: string,
  textTransform?: SupportedTransformationsType,
|};

type Props =
  | {|
      id: TranslationKeys,
      values?: Object,
      ...CommonProps,
    |}
  | {|
      passThrough: ?string | ?number, // may be undefined because of IDX output
      ...CommonProps,
    |};

/**
 * Regular translation using registered keys:
 *
 * ```js
 * <Translation id="SingleHotel.Rating.Stars" />
 * ```
 *
 * So called "pass through" translation (this props is not being translated
 * but passed through the component).
 *
 * ```js
 * <Translation passThrough="★★★★★" />
 * ```
 *
 * It's possible to apply text transformations in both cases:
 *
 * ```js
 * <Translation passThrough="★★★★★" textTransform="uppercase" />
 * <Translation id="SingleHotel.Rating.Stars" textTransform="uppercase" />
 * ```
 */
export default class Translation extends React.Component<Props> {
  // use this property to highlight translations for screenshoting or debugging
  highlightTranslations = false;

  render = () => {
    const containerStyle = this.highlightTranslations
      ? {
          backgroundColor: this.props.id ? 'orange' : 'pink',
        }
      : undefined;

    if (this.props.id === undefined) {
      // ID prop is missing so let's assume it's "pass through" translation
      // (property passThrough may be undefined)
      return (
        <Text style={containerStyle}>
          {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
          {CaseTransform(
            String(this.props.passThrough),
            this.props.textTransform,
          )}
        </Text>
      );
    }

    const values = this.props.values || {};
    return (
      <Text style={containerStyle}>
        {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
        {CaseTransform(
          replaceValues(vocabulary[this.props.id], values),
          this.props.textTransform,
        )}
      </Text>
    );
  };
}
