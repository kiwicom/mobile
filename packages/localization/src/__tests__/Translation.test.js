// @flow strict

import * as React from 'react';
import renderer from 'react-test-renderer';
import { translate } from '@kiwicom/rnmodules';

import Translation from '../Translation';
import { replaceValues } from '../TranslationHelpers';

const CancellableTranslation = require('../CancellableTranslation');

let Component;
beforeEach(() => {
  Component = new Translation({
    // $FlowExpectedError: we are not using real key for testing purposes
    id: 'test.key',
  });
});

const getComponent = (id: string = 'test.key') => {
  // $FlowExpectedError: we are not using real key for testing purposes
  return renderer.create(<Translation id={id} />);
};

it('should call getTranslationAsync with the key parameter', () => {
  getComponent();
  expect(translate).toBeCalledWith('mobile.test.key');
});

it('replaces parameter', () => {
  expect(
    replaceValues('Your flight __number__ departs now.', {
      number: 42,
    }),
  ).toBe('Your flight 42 departs now.');
});

it('replaces parameters', () => {
  expect(
    replaceValues(
      'Your flight __1_number__ departs from __2_airport__ gate __3_gate__.',
      {
        number: 42,
        airport: 'PRG',
        gate: 'C2',
      },
    ),
  ).toBe('Your flight 42 departs from PRG gate C2.');
});

it('works without parameters', () => {
  expect(replaceValues('Your flight departs now.')).toBe(
    'Your flight departs now.',
  );
});

it('works without value to replace', () => {
  expect(replaceValues('Your flight __number__ departs now.')).toBe(
    'Your flight ? departs now.',
  );
  expect(replaceValues('Your flight __number__ departs now.', {})).toBe(
    'Your flight ? departs now.',
  );
});

it('works with pass through translations', () => {
  const Component = new Translation({
    passThrough: 'This text should be returned as is.',
  });
  expect(Component.render()).toMatchSnapshot();
});

it('works with optional pass through parameter', () => {
  const Component = new Translation({
    passThrough: undefined,
  });
  expect(Component.render()).toMatchSnapshot();
});

it('works with pass through translations and text transformations', () => {
  const Component = new Translation({
    passThrough: 'This text should be returned in uppercase.',
    textTransform: 'uppercase',
  });
  expect(Component.render()).toMatchSnapshot();
});

it('does not call set state if promise is rejected', async () => {
  let originalFunction = CancellableTranslation.cancellableTranslation;
  // $FlowExpectedError: Intentionally overwriting function to test outcome
  CancellableTranslation.cancellableTranslation = jest.fn(() => ({
    promise: new Promise((resolve, reject) => {
      reject();
    }),
  }));

  jest.spyOn(Component, 'setState');
  await Component.setTranslatedString();

  expect(Component.setState).not.toHaveBeenCalled();
  // $FlowExpectedError: Intentionally resetting function
  CancellableTranslation.cancellableTranslation = originalFunction;
});

it('asks for new translation if id changes', () => {
  jest.spyOn(Component, 'setTranslatedString');
  Component.componentDidUpdate({ id: 'new.id' });

  expect(Component.setTranslatedString).toHaveBeenCalled();
});

it('does not ask for new translation if id does not change', () => {
  jest.spyOn(Component, 'setTranslatedString');
  Component.componentDidUpdate({ id: 'test.key' });

  expect(Component.setTranslatedString).not.toHaveBeenCalled();
});
