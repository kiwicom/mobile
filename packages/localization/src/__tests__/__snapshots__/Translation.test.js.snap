// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works with optional pass through parameter 1`] = `null`;

exports[`works with pass through translations 1`] = `
<Text
  expo={false}
  style={
    Array [
      undefined,
      Object {
        "fontFamily": "CircularPro-Book",
      },
    ]
  }
>
  This text should be returned as is.
</Text>
`;

exports[`works with pass through translations and text transformations 1`] = `
<Text
  expo={false}
  style={
    Array [
      undefined,
      Object {
        "fontFamily": "CircularPro-Book",
      },
    ]
  }
>
  THIS TEXT SHOULD BE RETURNED IN UPPERCASE.
</Text>
`;
